lib/api/statistics.py

```diff
@@ -0,0 +1,62 @@
+from datetime import datetime, timedelta
+
+from rest_framework import status
+from rest_framework import generics
+from django.db.models import Count, QuerySet
+from django.utils import timezone
+
+from exercise.submission_models import Submission
+from .serializers import StatisticsSerializer
+
+
+class BaseStatisticsView(generics.RetrieveAPIView):
+    """
+    Returns submission statistics for the entire system, over a given time window.
+
+    Returns the following attributes:
+
+    - `submission_count`: total number of submissions.
+    - `submitters`: number of users submitting.
+
+    Operations
+    ----------
+
+    `GET /statistics/`:
+        returns the statistics for the system.
+
+    - URL parameters:
+        - `endtime`: date and time in ISO 8601 format indicating the end point
+          of time window we are interested in. Default: now.
+        - `starttime`: date and time in ISO 8601 format indicating the start point
+          of time window we are interested in. Default: one day before endtime
+    """
+    serializer_class = StatisticsSerializer
+    endtime = timezone.now()
+    starttime = endtime - timedelta(days=1)
+
+    def get_queryset(self) -> QuerySet:
+        queryset = Submission.objects.all()
+
+        try:
+            endtime = self.request.query_params.get('endtime')
+            if endtime:
+                self.endtime = datetime.fromisoformat(endtime)
+
+            starttime = self.request.query_params.get('starttime')
+            if starttime:
+                self.starttime = datetime.fromisoformat(starttime)
+        except ValueError:
+            # Not sure what to do on invalid query params. For now just ignore them.
+            pass
```

Comment:
Yes, 400 is the right code, but my problem was how to generate that from get_queryset(), that should return a QuerySet. Should query_param parsing rather be done some place else?

Reply:
@PasiSa You raise an exception: `raise serializers.ValidationError(...)`

If you used a serializer for input validation, then the validation and its errors would be "hidden" inside the serializer class implementation. You would then create a serializer object and call its `is_valid()` method instead of raising exceptions manually. `StatisticsSerializer` could validate the date inputs if the other fields had the attribute `read_only=True`.

* https://www.django-rest-framework.org/api-guide/exceptions/#validationerror
* https://www.django-rest-framework.org/api-guide/fields/#read_only
* https://www.django-rest-framework.org/api-guide/serializers/#validation
* views have built-in exception handling: https://www.django-rest-framework.org/api-guide/views/#handle_exceptionself-exc

Reply:
Right. Indeed, this is now a much cleaner way.